app/scripts/lib/WeakRefObjectMap.ts

```diff
@@ -0,0 +1,214 @@
+type WeakRefObject<T extends Record<string, object>> = {
+  [P in keyof T]: WeakRef<T[P]>;
+};
+
+/**
+ * `WeakRefObjectMap` is a custom map-like structure designed to hold key-value pairs where the values are objects.
+ * Unlike a standard `Map`, this implementation stores each property of the value objects as weak references.
+ * This means that the properties of the objects are not prevented from being garbage collected when there are no other
+ * references to them outside of this map.
+ *
+ * It is important to note that while the map itself behaves similarly to a standard `Map`, the weak references apply
+ * to each property of the objects stored as values. This means that individual properties of these objects may become
+ * unavailable (i.e., garbage collected) independently of one another. Users of this map should be prepared to handle
+ * cases where a property's value has been collected and is therefore `undefined`.
+ *
+ * This class was implemented to help with memory management of network client proxies used by the SelectedNetworkController
+ * to keep per domain selected networks in sync. The properties of the NetworkClient object (provider and blockTracker) are weakly
+ * referenced so that they can be garbage collected if/when a dapp connection ends without effective cleanup.
+ */
+
+export class WeakRefObjectMap<T extends Record<string, object>>
+  implements Map<string, T>
+{
+  /**
+   * Internal map to store keys and their corresponding weakly referenced object values.
+   */
+  private map: Map<string, WeakRefObject<T>>;
+
+  constructor() {
+    this.map = new Map();
+  }
+
+  /**
+   * Associates a key with a value in the map. If the key already exists, its associated value is updated.
+   * The values are stored as weak references.
+   *
+   * @param key - The key under which to store the value.
+   * @param value - The value to store under the specified key. Must be an object.
+   * @returns The `WeakRefObjectMap` instance.
+   */
+  set(key: string, value: T): this {
+    const weakRefValue: WeakRefObject<T> = {} as WeakRefObject<T>;
+    for (const keyValue in value) {
+      if (!Object.hasOwn(value, keyValue)) {
+        continue;
+      }
+      const item: T[typeof keyValue] = value[keyValue];
+      if (typeof item === 'object' && item !== null) {
+        weakRefValue[keyValue] = new WeakRef(item);
+      } else {
+        console.warn(
+          `Property ${String(
+            keyValue,
+          )} is not an object and cannot be weakly referenced.`,
+        );
+      }
+    }
+    this.map.set(key, weakRefValue);
```

Comment:
I had an earlier comment about removing this type assertion in favour of an annotation. I see that that suggestion was merged, but the assertion was added back, so now we have both 😅.

I had misunderstood the need for the assertion in my original suggestion, I realize now why it was added. Using an annotation failed because TypeScript correctly identified that it was invalid, as the type is empty here. Really this is a _partial_ `WeakRefObject<T>`. It's not complete until after the loop.

This is similar to cases where we use `reduce` to build something up into a complete object; it's partial inside the `reduce` callback, but complete at the end. Yet TypeScript can't tell that it's complete, so we need to assert that it is.

The approach I usually use in cases like this is to declare the type as Partial to begin with, so that we have proper type checking inside the loop itself. Then at the point where we know it's complete, we can use an assertion to align the type with that. e.g.

```suggestion
    const weakRefValue: Partial<WeakRefObject<T>> = {};
    for (const keyValue in value) {
      if (!Object.hasOwn(value, keyValue)) {
        continue;
      }
      const item: T[typeof keyValue] = value[keyValue];
      if (typeof item === 'object' && item !== null) {
        weakRefValue[keyValue] = new WeakRef(item);
      } else {
        console.warn(
          `Property ${String(
            keyValue,
          )} is not an object and cannot be weakly referenced.`,
        );
      }
    }
    this.map.set(key, weakRefValue as WeakRefObject<T>);
```

Reply:
Done here: https://github.com/MetaMask/metamask-extension/pull/23525/commits/9592db0d6c35dc613a7bab5bd7408b0737ac5781